src/random.rs

```diff
@@ -0,0 +1,174 @@
+// random number generators (anything that implements SeedableRng is valid).
+struct RngHolder {
+    rng: Box<dyn Any>,
+}
+
+struct RngData {
+    base_seed: u64,
+    rng_holders: RefCell<HashMap<TypeId, RngHolder>>,
+}
+
+crate::context::define_data_plugin!(
+    RngPlugin,
+    RngData,
+    RngData {
+        base_seed: 0,
+        rng_holders: RefCell::new(HashMap::new()),
+    }
+);
+
+#[allow(clippy::module_name_repetitions)]
+pub trait RandomContext {
+    fn set_base_random_seed(&mut self, base_seed: u64);
+
+    fn get_rng<R: RngId>(&self) -> RefMut<'_, R::RngType>;
+}
+
+impl RandomContext for Context {
+    /// Initializes the `RngPlugin` data container to store rngs as well as a base
+    /// seed. Note that rngs are created lazily when `get_rng` is called.
+    fn set_base_random_seed(&mut self, base_seed: u64) {
+        let data_container = self.get_data_container_mut::<RngPlugin>();
+        data_container.base_seed = base_seed;
+
+        // Clear any existing Rngs to ensure they get re-seeded when `get_rng` is called
+        let mut rng_map = data_container.rng_holders.try_borrow_mut().unwrap();
+        rng_map.clear();
+    }
+
+    /// Gets a mutable reference to the random number generator associated with the given
+    /// `RngId`. If the Rng has not been used before, one will be created with the base seed
+    /// you defined in `set_base_random_seed`. Note that this will panic if `set_base_random_seed` was not called yet.
+    fn get_rng<R: RngId>(&self) -> RefMut<'_, R::RngType> {
+        let data_container = self
+            .get_data_container::<RngPlugin>()
+            .expect("You must initialize the random number generator with a base seed");
+
+        let rng_holders = data_container.rng_holders.try_borrow_mut().unwrap();
+
+        RefMut::map(rng_holders, |holders| {
+            holders
+                .entry(TypeId::of::<R>())
+                // Create a new rng holder if it doesn't exist yet
+                .or_insert_with(|| {
+                    let base_seed = data_container.base_seed;
+                    let seed_offset = fxhash::hash64(R::get_name());
+                    RngHolder {
+                        rng: Box::new(R::RngType::seed_from_u64(base_seed + seed_offset)),
+                    }
+                })
+                .rng
+                .downcast_mut::<R::RngType>()
+                .unwrap()
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::context::Context;
+    use crate::random::RandomContext;
+    use rand::RngCore;
+
+    define_rng!(FooRng);
+    define_rng!(BarRng);
+
+    #[test]
+    fn get_rng_basic() {
+        let mut context = Context::new();
+        context.set_base_random_seed(42);
+
+        let mut foo_rng = context.get_rng::<FooRng>();
+        assert_eq!(foo_rng.next_u64(), 5113542052170610017);
+        assert_eq!(foo_rng.next_u64(), 8640506012583485895);
+        assert_eq!(foo_rng.next_u64(), 16699691489468094833);
```

Comment:
Do you really want to hardcode these values? This will break if Rust changes the algorithm, which they say they might.

![image](https://github.com/user-attachments/assets/1aba7ff0-edbf-4d1e-b134-794c535ea187)